### PR TITLE
misc: Thread-safe cache rewrite

### PR DESCRIPTION
**Pull request by enwask**
Rewrites `memoized_meth` and `memoized_generator` for concurrency, as well as dropping the legacy `memoized_func` in favor of `functools.cache`. Also adds a global lock for the symbol cache allowing thread-safe `Symbol` construction and cache manipulation.

The `memoized_meth` decorator now stores one cache per thread, whereas `memoized_generator` stores a single cache for a given method (though there may still be misses if the cache is initialized concurrently). This means neither method cache has a call-once guarantee.

Memoized generators block for the initial call of the generator function and construct a thread-safe version of `itertools.tee`, which allows for concurrent iteration (but blocks when iterating elements that are not yet in the buffer). After the source generator is consumed, there is no blocking and subsequently threads can iterate the buffer in parallel (see `SafeTee` in `tools.memoization`).

### CODE REVIEW

File: devito/arch/archinfo.py [hunk 1]
```diff
@@ -41,7 +41,7 @@
            'PVC', 'INTELGPUMAX', 'MAX1100', 'MAX1550']
 
 
-@memoized_func
+@cache
```
Comment (FabioLuporini) [inline devito/arch/archinfo.py:44]: With `memoized_func` I know, IIRC, that the memoization cannot be evicted for the entire lifetime of the program. Is it still the same here? is this specific cache unbounded? otherwise, we may need an unbounded cache that we only surgically use for selected "system" functions, otherwise we would introduce a change in functionality

File: devito/tools/memoization.py [hunk 2]
```diff
@@ -1,135 +1,217 @@
...
 
-        code.activestate.com/recipes/577452-a-memoize-decorator-for-instance-methods/
-    """
+    def _get_cache(self, obj: InstanceType) -> dict[Hashable, ReturnType]:
+        """
+        Retrieves the thread-local cache for the given object instance, initializing
+        it if necessary.
+        """
+        # Try-catch is theoretically faster on the happy path
+        _local: local
```
Comment (FabioLuporini) [inline devito/tools/memoization.py:41]: this is improvable... :)

File: devito/tools/memoization.py [hunk 3]
```diff
@@ -1,135 +1,217 @@
...
+        try:
+            # Attempt to access the thread-local data
+            _local = obj._memoized_meth__local
 
-    def __init__(self, func):
-        self.func = func
+        # If the cache doesn't exist, initialize it
+        except AttributeError:
+            with self._lock:
```
Comment (FabioLuporini) [inline devito/tools/memoization.py:48]: if it's thread-local why would you need a lock?

File: devito/tools/memoization.py [hunk 4]
```diff
@@ -1,135 +1,217 @@
...
+
         return res
 
 
-class memoized_generator:
+# Describes the type of element yielded by a cached iterator
+YieldType = TypeVar('YieldType', covariant=True)
 
+
+class SafeTee(Iterator[YieldType]):
```
Comment (FabioLuporini) [inline devito/tools/memoization.py:86]: we could call this just `tee` and return python's `tee` (overriding `__new__`)  if mono thread

File: devito/types/basic.py [hunk 5]
```diff
@@ -559,24 +559,31 @@ def _cache_key(cls, *args, **kwargs):
...
+        obj = cls._cache_get(key)
         if obj is not None:
             return obj
 
-        # Not in cache. Create a new Symbol via sympy.Symbol
-        args = list(args)
-        name = kwargs.pop('name', None) or args.pop(0)
-        newobj = cls.__xnew__(cls, name, **assumptions)
+        # Lock against the symbol cache and double-check the cache
+        with CacheManager.lock():
```
Comment (FabioLuporini) [inline devito/types/basic.py:569]: I'm not comfortable with introducing thread-safety logic into the types hierarchy, because `__new__` gets overridden often, so this might create unexpected issues and subsequently burden on the developer. Why don't we move all this complexity into `types/caching`?

Secondly, if mono thread, it should somehow fallback to the previous "unlocked cache" data structure, unless you tell me the overhead of acquiring and releasing a lock with mono thread is practically zero

File: devito/types/caching.py [hunk 6]
```diff
@@ -76,8 +78,10 @@ def _cache_get(cls, key):
             obj = obj_cached()
             if obj is None:
                 # Cleanup _SymbolCache (though practically unnecessary)
-                # does not fail if it's already gone
-                _SymbolCache.pop(key, None)
+                with _cache_lock:
```
Comment (FabioLuporini) [inline devito/types/caching.py:81]: this code you're changing is monumentally critical, and at this stage I cannot claim these changes are safe yet

File: devito/types/caching.py [hunk 7]
```diff
@@ -196,16 +200,21 @@ def clear(cls, force=True):
...
 
-        for key in cache_copied:
-            obj = _SymbolCache.get(key)
-            if obj is None:
-                # deleted by another thread since we took the copy
-                continue
-            if obj() is None:
-                # (key could be removed in another thread since get() above)
-                _SymbolCache.pop(key, None)
+        for key, obj_cached in cache_copied.items():
```
Comment (FabioLuporini) [inline devito/types/caching.py:203]: this code you're changing is monumentally critical, and at this stage I cannot claim these changes are safe yet

**Review by FabioLuporini — COMMENTED** [review]
at a bare minimum we need a new `TestMultithreading `batch in `test_caching`

File: devito/tools/memoization.py [hunk 8]
```diff
@@ -1,135 +1,217 @@
...
-    If a memoized method is invoked directly on its class the result will not
-    be cached. Instead the method will be invoked like a static method: ::
 
-        class Obj:
-            @memoize
-            def add_to(self, arg):
-                return self + arg
-        Obj.add_to(1) # not enough arguments
-        Obj.add_to(1, 2) # returns 3, result is not cached
+    def __init__(self, meth: Callable[Concatenate[InstanceType, ParamsType],
```
Comment (EdCaunt) [inline devito/tools/memoization.py:22]: Tidy the typing up by putting `Callable[Concatenate[InstanceType, ParamsType], ReturnType]` on its own line as `methodtype` or similar

File: devito/tools/memoization.py [hunk 9]
```diff
@@ -1,135 +1,217 @@
...
+    def _get_cache(self, obj: InstanceType) -> dict[Hashable, ReturnType]:
+        """
+        Retrieves the thread-local cache for the given object instance, initializing
+        it if necessary.
+        """
+        # Try-catch is theoretically faster on the happy path
+        _local: local
+        try:
+            # Attempt to access the thread-local data
+            _local = obj._memoized_meth__local
```
Comment (EdCaunt) [inline devito/tools/memoization.py:44]: Maybe call it `on_local`? The double underscore in `obj._memoized_meth__local` is upsetting

File: devito/tools/memoization.py [hunk 10]
```diff
@@ -1,135 +1,217 @@
...
+        # If the cache doesn't exist, initialize it
         except AttributeError:
-            cache = obj.__cache_gen = {}
-        key = (self.func, args[1:], frozenset(kwargs.items()))
-        it = cache[key] if key in cache else self.func(*args, **kwargs)
-        cache[key], result = tee(it)
-        return result
+            with self._lock:
+                # Check again in case another thread initialized outside the lock
+                if not hasattr(obj, '_memoized_generator__cache'):
```
Comment (EdCaunt) [inline devito/tools/memoization.py:185]: Nitpick: I think the double underscore can be reduced to a single one in this method name

File: devito/tools/memoization.py [hunk 11]
```diff
@@ -1,135 +1,217 @@
...
+        key = (self._meth, args, frozenset(kwargs.items()))
+        try:
+            # Try to retrieve the cached value
+            res = cache[key]
+        except KeyError:
+            # If not cached, compute the value
+            source_iter = self._meth(obj, *args, **kwargs)
+            res = cache[key] = SafeTee(source_iter)
+
+        return res.tee()
```
Comment (EdCaunt) [inline devito/tools/memoization.py:209]: Isn't this essentially copying the `SafeTee` you have just created? Why do you need to do that?

File: devito/types/basic.py (hunk 5 above)
Comment (EdCaunt) [reply devito/types/basic.py:569]: Seconded on both of these points. Wrt the first, it is also worth considering that the integration of cache locks into the `__new__` effectively acts as a barrier to modifying and working on subclasses of `Symbol` (possibly preventing contributions), increases the chance for bugs to creep in though mistakes, and will generally uglify and complicate the codebase since this change will need propagating down into subclasses

File: devito/types/basic.py [hunk 12]
```diff
@@ -559,24 +559,31 @@ def _cache_key(cls, *args, **kwargs):
...
 
-        # Store new instance in symbol cache
-        Cached.__init__(newobj, key)
+            # Initialization
+            newobj._dtype = cls.__dtype_setup__(**kwargs)
+            newobj.__init_finalize__(name, *args, **kwargs)
 
-        return newobj
+            # Store new instance in symbol cache
+            Cached.__init__(newobj, key)
```
Comment (EdCaunt) [inline devito/types/basic.py:584]: This is a fairly fundamental class in Devito and will need substantial testing to be sure that these changes are not breaking any edge cases

File: tests/test_tools.py [hunk 13]
```diff
@@ -209,3 +211,156 @@ def __init__(self, value: int):
         # Cache should be cleared after Operator construction
         cache_size = Object._instance_cache.cache_info()[-1]
         assert cache_size == 0
+
+
+class TestMemoizedMethods:
```
Comment (EdCaunt) [inline tests/test_tools.py:216]: Do these tests get specifically run in the GIL-less CI?

Comment (EdCaunt) [reply tests/test_tools.py:216]: You should also probably have some tests to make sure this doesn't do anything weird when using `_rebuild`

File: devito/arch/archinfo.py (hunk 1 above)
Comment (enwask) [reply devito/arch/archinfo.py:44]: Yes, it's shorthand for `@lru_cache(maxsize=None)` which is unbounded

File: devito/tools/memoization.py (hunk 3 above)
Comment (enwask) [reply devito/tools/memoization.py:48]: The attribute itself contains thread-local data but it still needs to exist on the instance—I *think* it still needs to be initialized safely or one thread might erase existing thread local storage for another thread.

Although on second thought, maybe the cost of that occasional clash is worth getting rid of the lock

File: devito/types/caching.py (hunk 6 above)
Comment (enwask) [reply devito/types/caching.py:81]: Yeah, was going to try to figure out some reasonable tests for this

File: devito/tools/memoization.py (hunk 4 above)
Comment (enwask) [reply devito/tools/memoization.py:86]: Feels a little gross to be honest. What's the purpose of doing that if `SafeTee` isn't needed outside of memoized generators?

File: devito/tools/memoization.py (hunk 9 above)
Comment (enwask) [reply devito/tools/memoization.py:44]: The double underscore is emulating name mangling (like, when a class `Something` declares an attribute `__starting_with_double_underscores` it's mangled to `_Something__starting_with_double_underscores` to avoid clashing). Can rename it, but I did want it to be clear that this attribute is assigned by `memoized_meth` if someone happens to inspect a live object later and wonder what the heck `_on_local` comes from

File: devito/tools/memoization.py (hunk 10 above)
Comment (enwask) [reply devito/tools/memoization.py:185]: See reasoning above; this emulates name mangling but I guess just removing one of the underscores is not unreasonable

File: devito/tools/memoization.py (hunk 11 above)
Comment (enwask) [reply devito/tools/memoization.py:209]: Fair point, yeah we can just return the root `SafeTee` if we created it here

File: devito/tools/memoization.py [hunk 14]
```diff
@@ -1,135 +1,217 @@
...
-    be cached. Instead the method will be invoked like a static method: ::
 
-        class Obj:
-            @memoize
-            def add_to(self, arg):
-                return self + arg
-        Obj.add_to(1) # not enough arguments
-        Obj.add_to(1, 2) # returns 3, result is not cached
+    def __init__(self, meth: Callable[Concatenate[InstanceType, ParamsType],
+                                      ReturnType]) -> None:
```
Comment (JDBetteridge) [inline devito/tools/memoization.py:23]:
```suggestion
    def __init__(
        self,
        meth: Callable[Concatenate[InstanceType, ParamsType], ReturnType]
    ) -> None:
```

Comment (JDBetteridge) [reply devito/tools/memoization.py:23]: This would be my preferred style for all function signatures that are too long due to type hinting

File: devito/tools/memoization.py [hunk 15]
```diff
@@ -1,135 +1,217 @@
...
+        return partial(self, obj)
 
-        code.activestate.com/recipes/577452-a-memoize-decorator-for-instance-methods/
-    """
+    def _get_cache(self, obj: InstanceType) -> dict[Hashable, ReturnType]:
+        """
+        Retrieves the thread-local cache for the given object instance, initializing
+        it if necessary.
+        """
+        # Try-catch is theoretically faster on the happy path
```
Comment (JDBetteridge) [inline devito/tools/memoization.py:40]: Faster than what?

File: devito/tools/memoization.py [hunk 16]
```diff
@@ -1,135 +1,217 @@
...
+        """
+        Invokes the memoized method, caching the result if it hasn't been evaluated yet.
+        """
+        # Get the local cache for the object instance
+        cache = self._get_cache(obj)
+        key = (self._meth, args, frozenset(kwargs.items()))
         try:
+            # Try to retrieve the cached value
             res = cache[key]
         except KeyError:
```
[the remark this answers is not in the document]
Comment (JDBetteridge) [reply devito/tools/memoization.py:75]: After re-reading the AI, I might agree. Backtraces that encounter caches which use exception handling for flow control are painful to read

File: tests/test_tools.py [hunk 17]
```diff
@@ -209,3 +211,156 @@ def __init__(self, value: int):
...
+        Tests basic functionality of memoized_meth
+        """
+        class Object:
+            def __init__(self):
+                self.misses = 0
+
+            @memoized_meth
+            def compute(self, x):
+                self.misses += 1
+                return x * 2
```
Comment (JDBetteridge) [inline tests/test_tools.py:229]: Can these deeply nested objects be fixtures?

**Review by JDBetteridge — CHANGES_REQUESTED** [review]
I would like to see some more tests and maybe a bit more exposition. Maybe not a notebook, but a wiki page. Some reference for how all this is designed and supposed to work.

When I had to previously fix major bugs in a caching framework, most of what I was doing was fixing how people were incorrectly using the existing cache. There needs to be a readily available reference with examples, pitfalls and anti-patterns.

Most contributors (I expect) will have never had to deal with concurrent code execution or the principles that the whole paradigm is based on.

File: devito/tools/memoization.py (hunk 14 above)
Comment (enwask) [reply devito/tools/memoization.py:23]: Black style! Yeah I have been sorta torn between these two styles but if there's a preference I'll go with this

File: devito/tools/memoization.py (hunk 15 above)
Comment (enwask) [reply devito/tools/memoization.py:40]: Faster than two dictionary lookups

File: tests/test_tools.py (hunk 17 above)
Comment (enwask) [reply tests/test_tools.py:229]: Yeah I was thinking of factoring them out to fixtures, will do that when I revisit more cache testing

File: devito/tools/memoization.py (hunk 15 above)
Comment (FabioLuporini) [reply devito/tools/memoization.py:40]: add it maybe

File: devito/tools/memoization.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/tools/memoization.py:48]: I see now. 

Potentially easy workaround (brutal pseudocode just to give u the gist):

```
def  ThreadSafeCache(dict):

     # {thread_id -> {key -> value}}

     def __init__(.......):
           <this has to lead to a dict of dict somehow... perhaps subclass default dict... not sure>
           
      def __getitem__(self, k, v=None):
             return self[threadid][k] ...
```

This way you need no lock

I had to do something similar in the dark age here https://github.com/devitocodes/devito/blob/main/devito/tools/timing.py#L17 for slightly different reasons but still multi-threading related

File: devito/tools/memoization.py (hunk 4 above)
Comment (FabioLuporini) [reply devito/tools/memoization.py:86]: I mean, ultimately all I want is that single-threaded programs do not suffer from the inherent overheads of thread-safe data structures.

So, re-reading my comment, I'm guilty of mixing up syntactic sugar and suggestions; the suggestion would be falling back to plain tee if mono thread (does it matter? again, the mono thread program overhead when switching to thread-safe data structure is still unclear to mean -- and hopefully none!)

File: devito/tools/memoization.py [hunk 18]
```diff
@@ -1,135 +1,217 @@
...
+YieldType = TypeVar('YieldType', covariant=True)
 
+
+class SafeTee(Iterator[YieldType]):
     """
-    Decorator. Cache the return value of an instance generator method.
+    A thread-safe version of `itertools.tee` that allows multiple iterators to safely
+    share the same buffer.
+
+    In theory, this comes at a cost to performance of iterating elements that haven't
```
Comment (FabioLuporini) [inline devito/tools/memoization.py:91]: I started typing a long reply to this, but then I deleted it... I'll raise a question instead:

does this make data dependence analysis (Scope) run faster when executing with multiple threads, and what's the cutoff point ? I imagine mono thread will suffer from overhead

depending on the answer to this question, my next question would be: is this really the way we want to exploit multi-threading to parallelise data dependence analysis?

Generators are inherently sequential, and it feels like here we're forcing innatural behavior... won't all threads clash by attempting to yield contiguous elements back to back? do u see what I mean here?

Comment (FabioLuporini) [reply devito/tools/memoization.py:91]: I mean, part of the issue here is that we have to (want to) maintain the structure of the existing implementation based on generators -- an implementation that over the years has been extremely optimized for single threaded applications. If multi-threading is all we had to worry about, we might as we well forget about generators and use the thread pool to collectively construct the entire set of dependencies once and for all...

Comment (FabioLuporini) [reply devito/tools/memoization.py:91]: another thought spurred by ☝️ : 

we could have the thread pool construct a subset of Dependencies (e.g. each thread constructs K Dependencies at a time, possibly K=1) and then yield them one at a time

another orthogonal idea: we use N-1 threads to eagerly populate the Scope at instantiation time, so that when at the caller site we access any of the Scope attributes we find Dependencies already pre-built (rather than constructing them on the fly).

Comment (FabioLuporini) [reply devito/tools/memoization.py:91]: in this latter idea, we would be creating a producer-consumer pattern . The guy who instantiate the Scope is the consumer, and instantiation fires up the producer(s, ?) threads that are now in charge of populating the Scope, while the consumer runs its logic one element at a time

File: devito/tools/memoization.py (hunk 3 above)
Comment (enwask) [reply devito/tools/memoization.py:48]: Well then you need a lock for all accesses no? Or at least to place dictionaries in the top-level dict, since afaik that is not safe for concurrent modification

File: devito/tools/memoization.py (hunk 4 above)
Comment (enwask) [reply devito/tools/memoization.py:86]: Yeah, I mean with falling back—that's something I was going to go through and update for all of this code, to the extent that it's possible: avoid any locking overhead if the GIL is enabled (since presumably we won't be trying to multithread anything with the GIL as everything is basically interpreter bound)

Comment (enwask) [reply devito/tools/memoization.py:86]: The overhead is arguably negligible but definitely nonzero. I can try to do some profiling to see if it's worth eliminating when the GIL is enabled

File: devito/tools/memoization.py (hunk 18 above)
Comment (enwask) [reply devito/tools/memoization.py:91]: Yeah, this is something I was thinking about a lot when I was rewriting the generator cache. I wanted to see if eagerly evaluating them once might make more sense for multi-threaded use.

My concern was that lots of the higher-level opportunities for parallelism are (or contain) methods that heavily utilize `Scope`s, so there might be potential for a deadlock if a bunch of threads are waiting for a `Scope` to be populated but there aren't any workers available to act as producers. So this might need some more complex machinery like a work queue I think? And I know how you feel about complexity

File: devito/tools/memoization.py (hunk 14 above)
Comment (EdCaunt) [reply devito/tools/memoization.py:23]: I would personally just stick `Callable[Concatenate[InstanceType, ParamsType], ReturnType]` on its own line, like `MethodType = Callable[Concatenate[InstanceType, ParamsType], ReturnType]`
